### PR TITLE
[enocean] Added Eltako FPE1 and FPE2 contact sensors

### PR DESCRIPTION
**Pull request by henglert**
Added the Eltako FPE-1 and Eltako FPE-2 contact sensors

[enocean] Eltako FPE-1 and FPE-2 contact sensors use the EEP F6-10-00, but with different data interpretation than defined by enocean.

They send just one RPS data byte with value 0x00 for OPEN and 0x10 for CLOSED (FPE-1)  or 0x00 for CLOSED and 0x10 for OPEN (FPE-2). Since this does not comply to the enocean specification of 1111xxxx for CLOSED and 11x0xxxx for OPEN, I implemented two new contact types for these devices.

### CODE REVIEW

**Comment by henglert** [comment]
Struggling with github pull requests a little bit.

**Comment by wborn** [comment]
Don't mind the failing Jenkins build. It also fails for other PR's!

**Comment by henglert** [comment]
Just a comment: Wouldn't it be easier to maintain if the (huge) openhab-addons git repository would be split up into different smaller bindings repositories?

**Comment by wborn** [comment]
I don't mind the size because I've worked on much bigger code repos. :wink:  IMHO this way it has less overhead and it's easier to maintain. It's also easier for users because they just have to deal with fewer issue trackers. Because all add-on code is in this repo there's more people reviewing code and triaging issues.

**Comment by henglert** [comment]
Sorry, can I ask you (or maybe better fruggy83) one more question: fruggy83 still has his own openocean binding repository with some extensions that did not yet go into the enocean binding, e.g. environmental sensor Eltako .  I merged that sensor Eltako FWS61-24V DC from his repository (old build system) into the enocean binding and it works great. Wouldn't it make sense to work on just one repository for the enocean/openocean binding, not two? So there would be more testers (and the testers wouldtn't have to merge the code each by themselves).

**Comment by wborn** [comment]
That's a good question that probably only @fruggy83 can answer. Maybe you can also add these reasons and OH differences to the README.md in your [openocean](https://github.com/fruggy83/openocean) repo?

I can imagine that when there are many PRs for an add-on (or it has an incompatible license) it's easier to maintain (or required) to create a separate add-on repo for it. I.e. like the zwave/zigbee/bacnet add-ons.

File: bundles/org.openhab.binding.enocean/src/main/java/org/openhab/binding/enocean/internal/eep/EEPFactory.java [hunk 1]
```diff
@@ -97,7 +99,21 @@ public static EEP buildEEPFromTeachInERP1(ERP1Message msg) {
                     }
                 } catch (Exception e) {
                 }
+                try {
+                    EEP result = new F6_10_00_EltakoFPE1(msg);
+                    if (result.isValid()) {
```
Comment (fruggy83) [inline bundles/org.openhab.binding.enocean/src/main/java/org/openhab/binding/enocean/internal/eep/EEPFactory.java]: Do we have a more reliable chance to detect these two new EEPs? Can we distinguish FPE1 messages from FPE2 messages? I think we should at least override the validateData function of these EEPs and check if we received a 0x10 or 0x00. You could also change the check order of your EEP and F6_10_01. Your EEP should never be recognized as F6_10_01 as I check [bit 6, 3 and 2](https://github.com/openhab/openhab2-addons/blob/ecc69c76dc6fb023e56eca7b67b40e61ae6ef3d0/bundles/org.openhab.binding.enocean/src/main/java/org/openhab/binding/enocean/internal/eep/F6_10/F6_10_01.java#L79). Currently we are no longer able to autodectect a F6_10_01 EEP (window handle).

**Comment by fruggy83** [comment]
Thanks a lot for your work 👍  Your changes look really good. I have just a comment/question regarding the detection of these new EEPs.

Regarding my own (op)enocean binding repo:
I am currently using my repo as a playground repo. I introduce new features in my repo first, publish a new release and let it test by the users. After a successful test I create a PR in the official repo. I do not own the devices of all implemented EEP (like the environmental sensor). So I have to implement them just with the help of the specs and without the chance to test them by myself. With the help of my repo the user testing is a lot more easier and faster. Furthermore I know a lot of users who do not want to update their complete openhab system (switch to snapshot, risk of messing up the system) to test a new version of the enocean binidng. @wborn Correct me if I am wrong, but currently there is no chance to just update a single binding (or switch to the snapshot version) without touching the other bindings or the whole system?

To sum up, sooner or later all features implemented in my repo will find their way back into the openhab addons repo.

Best regards
Daniel

File: bundles/org.openhab.binding.enocean/src/main/java/org/openhab/binding/enocean/internal/eep/EEPFactory.java (hunk 1 above)
Comment (henglert) [reply bundles/org.openhab.binding.enocean/src/main/java/org/openhab/binding/enocean/internal/eep/EEPFactory.java]: You're right. Sorry, I'm starting to understand your code, but don't have the full view yet. The only FPE-1 and FPE-2 documentation I have is from the web, and according to that, they only send one byte 0x00 and 0x10 for Open/Closed or vice versa, so there's no change to distinguish FPE-1 from FPE-2 messages. So that distinction should be made by configuration. I don't own a FPE-2, only FPE-1's so I have to trust on Eltako's documentation about only swapping the OPEN/CLOSED values.
If I validate the FPE-X-data by the pattern 000X0000, that should work, what do you think?
Implementation will last a few days, since I don't have experience with the evaluation of configuration options.

**Comment by wborn** [comment]
> but currently there is no chance to just update a single binding (or switch to the snapshot version) without touching the other bindings or the whole system?

Most users update single bindings to a snapshot version by uninstalling the binding in Paper UI and then adding the snapshot JAR to the addons directory.

Switching back is done in the reverse order by removing the JAR from the addons directory and installing the binding via Paper UI again.

If you know the bundle ID and bundle URLs you can also switch versions very easily using the Karaf console with:

`bundle:update <bundle-id> <download-url>`

Example:

```
openhab> bundle:list | grep EnOcean                                                                                                                                                                                                    
251 │ Active   │  80 │ 2.4.0                  │ EnOcean Binding
openhab> bundle:update 251 https://openhab.jfrog.io/openhab/online-repo-snapshot/2.5/org/openhab/addons/bundles/org.openhab.binding.enocean/2.5.0-SNAPSHOT/org.openhab.binding.enocean-2.5.0-SNAPSHOT.jar
openhab> bundle:list | grep EnOcean
251 │ Active   │  80 │ 2.5.0.201905290421     │ openHAB Add-ons :: Bundles :: EnOcean Binding
openhab> bundle:update 251 https://dl.bintray.com/openhab/mvn/online-repo/2.4/org/openhab/binding/org.openhab.binding.enocean/2.4.0/org.openhab.binding.enocean-2.4.0.jar
openhab> bundle:list | grep EnOcean
251 │ Active   │  80 │ 2.4.0                  │ EnOcean Binding
```

File: bundles/org.openhab.binding.enocean/src/main/java/org/openhab/binding/enocean/internal/eep/EEPFactory.java (hunk 1 above)
Comment (fruggy83) [reply bundles/org.openhab.binding.enocean/src/main/java/org/openhab/binding/enocean/internal/eep/EEPFactory.java]: I also googled a bit and could not find any way to distinguish FPE-1 from FPE-2 messages too. So I would suggest the following:
 * add the validation pattern 0b000X0000
 * just autodetect either FPE-1 or FPE-2, the other one must be manually reconfigured by choosing the right receiving EEP
I would not add any more configuration options and just keep your two EEP implementations.
Currently RPS autodetection is pure guessing. I think I have to improve it a little bit. Could you please post the raw message of your contact sensor for closed and open?

Comment (henglert) [reply bundles/org.openhab.binding.enocean/src/main/java/org/openhab/binding/enocean/internal/eep/EEPFactory.java]: Your post came just a little bit too late for my commit. Here are the RAW messages:
OPEN: F600XXXXXXXX20
CLOSED: F610XXXXXXXX20
X-es are the 4 byte enocean ID of the sensor.
In my current commit I added a boolean configuration option for contact channel named "inverted" which swaps OPEN/CLOSED and defaults to 'false'. Maybe this could be useful also for other contact sensors? You decide, I can switch back.

Comment (fruggy83) [reply bundles/org.openhab.binding.enocean/src/main/java/org/openhab/binding/enocean/internal/eep/EEPFactory.java]: Thanks for your changes. That channel config parameter is really useful 👍 . However could you do me favor and check this paramter in EEP A5-14, D5-00 and F6-10, as we use the same contact channel there too. Furthermore it would be nice if you could state this new option in readme.md too. Thanks a lot in advance.

Comment (henglert) [reply bundles/org.openhab.binding.enocean/src/main/java/org/openhab/binding/enocean/internal/eep/EEPFactory.java]: I just returned from vacation and will start the tasks during this week.

**Comment by henglert** [comment]
I can't see the problem with the travis build (Compile Error with `convertToStateImpl`). I don't get this error when building and there is the `@Override` just before the function.

File: bundles/org.openhab.binding.enocean/src/main/java/org/openhab/binding/enocean/internal/eep/F6_10/F6_10_00_EltakoFPE.java [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+    public F6_10_00_EltakoFPE() {
+        super();
+    }
+
+    public F6_10_00_EltakoFPE(ERP1Message packet) {
+        super(packet);
+    }
+
+    @Override
+    protected State convertToStateImpl(String channelId, String channelTypeId, State currentState, Configuration config) {
```
Comment (J-N-K) [inline bundles/org.openhab.binding.enocean/src/main/java/org/openhab/binding/enocean/internal/eep/F6_10/F6_10_00_EltakoFPE.java]:
```suggestion
    protected State convertToStateImpl(String channelId, String channelTypeId, Function<String, State> getCurrentStateFunc, Configuration config) {
```

Comment (J-N-K) [reply bundles/org.openhab.binding.enocean/src/main/java/org/openhab/binding/enocean/internal/eep/F6_10/F6_10_00_EltakoFPE.java]: That should fix the compile error.

Comment (henglert) [reply bundles/org.openhab.binding.enocean/src/main/java/org/openhab/binding/enocean/internal/eep/F6_10/F6_10_00_EltakoFPE.java]: Then I get the compile error `[...]F6_10_00_EltakoFPE.java:[43,1397] Function cannot be resolved to a type`

Comment (J-N-K) [reply bundles/org.openhab.binding.enocean/src/main/java/org/openhab/binding/enocean/internal/eep/F6_10/F6_10_00_EltakoFPE.java]: I think you need to add some imports, too. Eclipse does that automatically on save. Did you rebase on master befor changing that? Otherwise you'll probably get another error.

**Comment by J-N-K** [comment]
Sign-Offs ok.

**Comment by J-N-K** [comment]
Jenkins fails with OoM.
